usr/opt/OSC200/usr/share/man/man3/puts.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: puts.3,v $ $Revision: 3.2.5.2 $ (DEC) $Date: 1993/06/25 15:59:02 $
.\"
.\" (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
.\" ALL RIGHTS RESERVED 
.\" OSF/1 1.2
.\"
.\"
.TH puts 3
.SH NAME
.PP
\*Lputs\*O, \*Lfputs\*O \-  Writes a string to a stream
.SH LIBRARY
.PP
Standard C Library (\*Llibc.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <stdio.h>\*O
.PP
\*Lint puts(
.nL
        const char *\*Vstring\*L);\*O
.PP
\*Lint fputs(
.nL
        const char *\*Vstring\*L,
.nL
        FILE *\*Vstream\*L);\*O
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vstring\*O"
Points to a string to be written to output.
.LI "\*Vstream\*O"
Points to the \*LFILE\*O structure of an open file.
.LE
.SH DESCRIPTION
.PP
.iX "puts function"
.iX "fputs function"
.iX "writing out a string"
.iX "strings" "writing out"
.iX "IO functions" "standard"
The \*Lputs(\|)\*O function writes the null-terminated string
pointed to by the \*Vstring\*O parameter, followed by a newline
character, to the standard output stream, \*Lstdout\*O.
.PP
The \*Lfputs(\|)\*O function writes the null-terminated string
pointed to by the \*Vstring\*O parameter to the output stream specified
by the \*Vstream\*O parameter. The \*Lfputs(\|)\*O function does
not append a newline character.
.PP
Neither function writes the terminating null byte.
.PP
The \*Lst_ctime\*O and \*Lst_mtime\*O
fields of the file are marked for update between the successful
execution of the \*Lputs(\|)\*O or \*Lfputs(\|)\*O function
and the next successful completion of a call to one of the following:
.ML
.LI
An \*Lfflush(\|)\*O or \*Lfclose(\|)\*O function on the same stream
.LI
The \*Lexit(\|)\*O or \*Labort(\|)\*O function
.LE
.SH NOTES
.PP
.VL 1.5i
.LI "\*LAES Support Level:\*O"
Full use.
.LE
.SH RETURN VALUES
.PP
Upon successful completion, the \*Lputs(\|)\*O and \*Lfputs(\|)\*O
functions return the number of characters written. Both subroutines
return \*LEOF\*O on an error.
.SH ERRORS
.PP
The \*Lputs(\|)\*O and \*Lfputs(\|)\*O functions fail under
either of the following conditions:
.ML
.LI
The \*Vstream\*O is unbuffered.
.LI
The \*Vstream\*O's
buffer needed to be flushed, the function call caused an underlying
\*Lwrite(\|)\*O or \*Llseek(\|)\*O to be invoked, and this
underlying operation fails with incomplete output.
.LE
.PP
In addition, if any of the following conditions occur, the
\*Lputs(\|)\*O or \*Lfputs(\|)\*O function sets
\*Lerrno\*O to the corresponding value.
.VL 1i
.LI "[EAGAIN]"
The \*LO_NONBLOCK\*O flag is set for the file descriptor
underlying \*Vstream\*O and the process would be delayed in the write operation.
.LI "[EBADF]"
The file descriptor underlying \*Vstream\*O
is not a valid file descriptor open for writing.
.LI "[EFBIG]"
An attempt was made to write to a file that exceeds the
process's file size limit or the maximum file size.
.LI "[EINTR]"
The read operation was interrupted by a signal that was caught,
and no data was transferred.
.LI "[EIO]"
The implementation supports job control; the process is a member
of a background process group attempting to write to its
controlling terminal; \*LTOSTOP\*O is set; the process is neither ignoring nor
blocking \*LSIGTTOU\*O, and the process group of the process is
orphaned.
This error may also be returned under implementation-defined conditions.
.LI "[ENOSPC]"
There was no free space remaining on the device containing the file.
.LI "[EPIPE]"
An attempt was made to write to a pipe or FIFO that is not open
for reading by any process.
A \*LSIGPIPE\*O signal will also be sent to the process.
.LE
.SH RELATED INFORMATION
.PP
Functions:
\*Lfgetws\*O(3),
\*Lfputws\*O(3),
\*Lgets\*O(3),
\*Lprintf\*O(3),
\*Lputc\*O(3),
\*Lputwc\*O(3).
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/puts.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: puts.c,v $ $Revision: 4.3.6.2 $ (DEC) $Date: 1993/06/07 21:19:41 $";
#endif 
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
#if !defined(lint) && !defined(_NOIDENT)

#endif
/*
 * FUNCTIONS: puts 
 *
 * (C) COPYRIGHT International Business Machines Corp. 1985, 1989 
 * All Rights Reserved
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 * puts.c	1.10  com/lib/c/io,3.1,8943 9/9/89 13:30:28
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#endif
#include <stdio.h>
#include "stdiom.h"
#include "ts_supp.h"

#ifdef	_THREAD_SAFE
#include "stdio_lock.h"
#endif	/* _THREAD_SAFE */

extern char *memccpy();

/*                                                                   
 * FUNCTION:	The puts function writes the string pointed to by s to the
 *		stream pointed to by stdout, and appends a new-line character
 *		to the output.  The terminating character is not written.
 *
 * 		This version reads directly from the buffer rather than
 *		looping on putc.  Ptr args aren't checked for NULL because
 *		the program would be a catastrophic mess anyway.  Better
 *		to abort than just to return NULL.
 *
 * RETURN VALUE DESCRIPTION:	
 *		The puts function returns EOF if a write error occurs;
 *		otherwise it returns a nonnegative value
 *
 */  

int 	
puts(const char *s)
{
	char	*p;
	int	ndone = 0, n;
	unsigned char	*cs, *bufend;
	TS_FDECLARELOCK(filelock)

	TS_FLOCK(filelock, stdout);

	if (_WRTCHK(stdout)) {
		TS_FUNLOCK(filelock);
		return (EOF);
	}

	bufend = _bufend(stdout);

	for ( ; ; s += n) {
		while ((n = bufend - (cs = stdout->_ptr)) <= 0) /* full buf */
			if (_xflsbuf(stdout) == EOF) {
				TS_FUNLOCK(filelock);
				return(EOF);
			}
		if ((p = memccpy((char *) cs, s, '\0', n)) != NULL)
			n = p - (char *) cs;
		stdout->_cnt -= n;
		stdout->_ptr += n;
		_BUFSYNC(stdout);
		ndone += n;
		if (p != NULL) {
			stdout->_ptr[-1] = '\n'; /* overwrite '\0' with '\n' */
			if (stdout->_flag & (_IONBF | _IOLBF)) /* flush line */
				if (_xflsbuf(stdout) == EOF) {
					TS_FUNLOCK(filelock);
					return(EOF);
				}
			TS_FUNLOCK(filelock);
			return(ndone);
		}
	}
}
